subprojects/core/src/main/java/org/gradle/util/internal/SimpleMapInterner.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.util.internal;
+
+import com.google.common.collect.Interner;
+import com.google.common.collect.Maps;
+
+import java.util.Map;
+
+public class SimpleMapInterner implements Interner<String> {
```

Comment:
Why not use the StringInterner we already have?

Reply:
Because it's too complex for what we want to do: the weak interner is slow, and the strong interner is slower.

Reply:
Then why are we still using those? We should have one way of interning stuff.

Reply:
Because the Guava interner we use is the weak one, which is global scope. I can afford using a strong interner because it's build scoped.

Reply:
Yeah but why not use `Interners.strongInterner()`?

Reply:
As I said if I use `strongInterner` then it's slower than my own interner, because Guava has more logic (like synchronizing access and guaranteed interning, while I don't care if 2 strings are not interned if by mischance the same string is interned concurrently by 2 threads).

Reply:
Got it, thanks.

Reply:
So to summarize:

- global scope interner uses Guava's weak interner, because it shouldn't leak memory. It has a cost, interning is pretty slow
- build scope uses my `SimpleMapInterner`, because I don't need the same guarantees as Guava, which is _real_ interning. It's ok for me to miss a couple of internings by accident. This interner is a strong interner, so leaks memory but it's ok since it's build scoped.
- we don't use Guava's strong interner because we don't have a good use case for it which is not solved by the above